File: python/src/bsb/telemetry.py
"""Bounded native telemetry exporters; transport work never runs in application log calls."""
import asyncio
from datetime import datetime
import json
from queue import Empty, Full, Queue
import sys
import time
from urllib.error import HTTPError, URLError

from .base import BSBObservable
from .http import json_request, origin
from .logging import LEVELS, redact
from .schema import av


def fields(endpoint):
    return {
        "endpoint": av.string().min_length(1).default(endpoint), "serviceName": av.string().min_length(1).default("bsb-service"),
        "serviceVersion": av.optional(av.string()), "headers": av.record(av.string()).default({}).describe("HTTP headers", sensitive=True),
        "resourceAttributes": av.record(av.string()).default({}),
        "flushIntervalMs": av.int32().min(100).max(60000).default(5000), "maxBatchSize": av.int32().min(1).max(4096).default(512),
        "samplingRate": av.number().min(0).max(1).default(1),
        "logs": av.bool_().default(True), "metrics": av.bool_().default(True), "traces": av.bool_().default(True),
        "level": av.enum_(LEVELS).default("info"), "redact": av.array(av.string().min_length(1)).default([]),
    }


def value(item):
    if isinstance(item, bool): return {"boolValue": item}
    if isinstance(item, int) and -(2**63) <= item < 2**63: return {"intValue": str(item)}
    if isinstance(item, float): return {"doubleValue": item}
    if isinstance(item, dict): return {"kvlistValue": {"values": attributes(item)}}
    if isinstance(item, (list, tuple)): return {"arrayValue": {"values": [value(part) for part in item]}}
    return {"stringValue": "null" if item is None else str(item)}


def attributes(items):
    return [{"key": str(key), "value": value(item)} for key, item in items.items()]


def resource(config):
    data = {**config["resourceAttributes"], "service.name": config["serviceName"]}
    if config.get("serviceVersion"):
        data["service.version"] = config["serviceVersion"]
    return {"attributes": attributes(data)}


def otlp(signal, entries, config):
    scopes = {}
    for entry in entries:
        plugin = entry.get("pluginName", entry.get("plugin", "bsb"))
        scopes.setdefault(plugin, []).append(entry)
    result = []
    for plugin, batch in scopes.items():
        scope = {"scope": {"name": plugin}}
        if signal == "traces":
            scope["spans"] = [{"traceId": item["trace"]["t"], "spanId": item["trace"]["s"],
                **({"parentSpanId": item["parentSpanId"]} if item.get("parentSpanId") else {}),
                "name": item["name"], "kind": 1, "startTimeUnixNano": str(item["startedNs"]),
                "endTimeUnixNano": str(item["startedNs"] + item["durationNs"]), "attributes": attributes(item["attributes"]),
                "status": {"code": 2, "message": item["error"]} if item.get("error") else {"code": 0}} for item in batch]
        elif signal == "logs":
            scope["logRecords"] = [{"timeUnixNano": str(int(datetime.fromisoformat(item["timestamp"]).timestamp() * 1_000_000_000)),
                "severityNumber": [1, 5, 9, 13, 17, 21][LEVELS.index(item["level"])], "severityText": item["level"].upper(),
                "body": value(item["message"]), "traceId": item["traceId"], "spanId": item["spanId"],
                "attributes": attributes(item.get("meta", {}))} for item in batch]
        else:
            metrics = {}
            # Cumulative instruments only need the newest observation of each series in this batch.
            series = {(item["name"], tuple(sorted(item["labels"].items()))): item for item in batch}
            for item in series.values():
                metric = metrics.setdefault(item["name"], {"name": item["name"], "description": item["description"], "unit": item["unit"]})
                point = {"startTimeUnixNano": str(item["startedNs"]), "timeUnixNano": str(item["timestampNs"]), "attributes": attributes(item["labels"])}
                if item["kind"] == "histogram":
                    point.update(count=str(item["count"]), sum=item["sum"], min=item["min"], max=item["max"], bucketCounts=[str(item["count"])], explicitBounds=[])
                    metric.setdefault("histogram", {"aggregationTemporality": 2, "dataPoints": []})["dataPoints"].append(point)
                else:
                    point["asDouble"] = item["value"]
                    data = metric.setdefault("sum" if item["kind"] == "counter" else "gauge", {"dataPoints": []})
                    if item["kind"] == "counter":
                        data.update(aggregationTemporality=2, isMonotonic=True)
                    data["dataPoints"].append(point)
            scope["metrics"] = list(metrics.values())
        result.append(scope)
    name = {"traces": "Spans", "logs": "Logs", "metrics": "Metrics"}[signal]
    return {"resource" + name: [{"resource": resource(config), "scope" + name: result}]}


def zipkin(entries, config):
    return [{"traceId": item["trace"]["t"], "id": item["trace"]["s"],
        **({"parentId": item["parentSpanId"]} if item.get("parentSpanId") else {}),
        "name": item["name"], "timestamp": item["startedNs"] // 1000, "duration": max(1, item["durationNs"] // 1000),
        "localEndpoint": {"serviceName": config["serviceName"]},
        "tags": {**{key: str(item) for key, item in item["attributes"].items()}, **({"error": item["error"]} if item.get("error") else {})}}
        for item in entries]


def post(url, body, headers):
    if len(json.dumps(body, allow_nan=False).encode()) > 64 * 1024 * 1024:
        raise ValueError("Telemetry request exceeds 64 MiB")
    for attempt in range(3):
        try:
            result = json_request("POST", url, body=body, headers=headers)
            if isinstance(result, dict):
                partial = result.get("partialSuccess", {})
                if any(int(partial.get(key, 0)) for key in ("rejectedLogRecords", "rejectedSpans", "rejectedDataPoints")) or partial.get("errorMessage"):
                    raise ValueError("Collector reported partial rejection")
                if result.get("failed", 0):
                    raise ValueError("Collector rejected events")
            return
        except HTTPError as error:
            if error.code not in (429, 502, 503, 504) or attempt == 2:
                raise
            retry = error.headers.get("Retry-After", "")
            delay = min(float(retry), 5) if retry.isdecimal() else .25 * (attempt + 1)
        except (URLError, TimeoutError, ConnectionError):
            if attempt == 2:
                raise
            delay = .25 * (attempt + 1)
        time.sleep(delay)


class BufferedTelemetry(BSBObservable):
    signals = ("logs", "metrics", "traces")

    def __init__(self, ctor):
        super().__init__(ctor)
        self._queue = Queue(4096)
        self._closed = False
        self._dropped = 0
        self._worker = None
        self._stop = asyncio.Event()

    def enqueue(self, signal, entry):
        if self._closed or signal not in self.signals or not self.config.get(signal, True):
            return
        if signal == "logs" and LEVELS.index(entry["level"]) < LEVELS.index(self.config.get("level", "info")):
            return
        if signal == "traces" and int(entry["trace"]["t"][:8], 16) / 2**32 >= self.config.get("samplingRate", 1):
            return
        item = {**redact(entry, self.config.get("redact", [])), "signal": signal}
        try:
            self._queue.put_nowait(item)
        except Full:
            self._dropped += 1
            if self._dropped % 1000 == 1:
                print(f"[{self.plugin_name}] Telemetry queue full; dropped {self._dropped} entries", file=sys.stderr)

    def emit_log(self, entry): self.enqueue("logs", entry)
    def emit_metric(self, entry): self.enqueue("metrics", entry)
    def emit_span(self, entry): self.enqueue("traces", entry)

    async def run(self, trace):
        if self._worker is None:
            self._worker = asyncio.create_task(self.pump())

    async def pump(self):
        while True:
            try:
                await asyncio.wait_for(self._stop.wait(), self.config.get("flushIntervalMs", 5000) / 1000)
            except TimeoutError:
                pass
            while not self._queue.empty():
                batch = []
                for _ in range(self.config.get("maxBatchSize", 512)):
                    try:
                        batch.append(self._queue.get_nowait())
                    except Empty:
                        break
                try:
                    await self.export(batch)
                except Exception as error:
                    print(f"[{self.plugin_name}] Export failed ({len(batch)} entries): {type(error).__name__}", file=sys.stderr)
            if self._closed:
                return

    async def dispose(self):
        if self._closed:
            return
        self._closed = True
        self._stop.set()
        if self._worker is None:
            self._worker = asyncio.create_task(self.pump())
        try:
            await asyncio.wait_for(self._worker, 10)
        except TimeoutError:
            print(f"[{self.plugin_name}] Telemetry shutdown deadline reached", file=sys.stderr)


class HttpTelemetry(BufferedTelemetry):
    async def init(self, trace):
        origin(self.config["endpoint"], allow_http=True)

    def endpoint(self, path):
        return self.config["endpoint"].rstrip("/") + "/" + path.lstrip("/")

    async def send(self, url, body, headers=None):
        await asyncio.to_thread(post, url, body, self.config["headers"] if headers is None else headers)

File: rust/src/telemetry.rs
use crate::{http, observable::Observer};
use anyhow::{Result, ensure};
use async_trait::async_trait;
use reqwest::header::{HeaderMap, HeaderName, HeaderValue};
use serde::Deserialize;
use serde_json::{Value, json};
use std::{
    collections::BTreeMap,
    sync::{
        Arc, Mutex,
        atomic::{AtomicBool, AtomicU64, Ordering},
    },
    time::Duration,
};
use tokio::sync::{Mutex as AsyncMutex, mpsc};
use tokio_util::sync::CancellationToken;

#[derive(Clone, Deserialize)]
#[serde(default, rename_all = "camelCase", deny_unknown_fields)]
pub(super) struct Options {
    pub mode: String,
    pub level: String,
    pub redact: Vec<String>,
    pub base: Value,
    pub pretty_print: bool,
    pub path: String,
    pub file_path: String,
    pub max_bytes: u64,
    pub max_files: usize,
    pub interval: String,
    pub compress: bool,
    pub endpoint: String,
    pub service_name: String,
    pub service_version: String,
    pub headers: BTreeMap<String, String>,
    pub resource_attributes: BTreeMap<String, String>,
    pub flush_interval_ms: u64,
    pub max_batch_size: usize,
    pub sampling_rate: f64,
    pub logs: bool,
    pub metrics: bool,
    pub traces: bool,
    pub token: String,
    pub dataset: String,
    pub org_id: String,
    pub allow_insecure_http: bool,
    pub host: String,
    pub port: u16,
    pub protocol: String,
    pub facility: Value,
    pub hostname: String,
    pub app_name: String,
    pub rfc: String,
    pub framing: String,
    pub ca_certificate_path: String,
    pub client_certificate_path: String,
    pub client_key_path: String,
    pub http_endpoint: String,
    pub additional_fields: Value,
}
impl Default for Options {
    fn default() -> Self {
        Self {
            mode: "production".into(),
            level: "info".into(),
            redact: vec![],
            base: json!({}),
            pretty_print: false,
            path: "logs/application.log".into(),
            file_path: String::new(),
            max_bytes: 10 * 1024 * 1024,
            max_files: 7,
            interval: "daily".into(),
            compress: true,
            endpoint: String::new(),
            service_name: "bsb-service".into(),
            service_version: String::new(),
            headers: BTreeMap::new(),
            resource_attributes: BTreeMap::new(),
            flush_interval_ms: 5000,
            max_batch_size: 512,
            sampling_rate: 1.0,
            logs: true,
            metrics: true,
            traces: true,
            token: String::new(),
            dataset: "bsb-logs".into(),
            org_id: String::new(),
            allow_insecure_http: false,
            host: "localhost".into(),
            port: 514,
            protocol: "udp".into(),
            facility: json!(16),
            hostname: std::env::var("HOSTNAME").unwrap_or_else(|_| "localhost".into()),
            app_name: "bsb-app".into(),
            rfc: "5424".into(),
            framing: "newline".into(),
            ca_certificate_path: String::new(),
            client_certificate_path: String::new(),
            client_key_path: String::new(),
            http_endpoint: String::new(),
            additional_fields: json!({}),
        }
    }
}
pub(super) fn level(value: &str) -> Option<u8> {
    match value {
        "trace" => Some(0),
        "debug" => Some(1),
        "info" => Some(2),
        "warn" => Some(3),
        "error" => Some(4),
        "fatal" => Some(5),
        _ => None,
    }
}
pub struct Native {
    kind: String,
    options: Options,
    sender: mpsc::Sender<Value>,
    stop: CancellationToken,
    closed: AtomicBool,
    dropped: AtomicU64,
    worker: AsyncMutex<Option<tokio::task::JoinHandle<()>>>,
    file: Option<Mutex<crate::telemetry_file::RotatingFile>>,
}
impl Native {
    pub async fn new(kind: &str, raw: Value) -> Result<Arc<Self>> {
        ensure!(
            [
                "observable-default",
                "observable-logging-file",
                "observable-pino",
                "observable-winston",
                "observable-opentelemetry",
                "observable-axiom",
                "observable-zipkin",
                "observable-graylog",
                "observable-syslog"
            ]
            .contains(&kind),
            "unknown native observable plugin"
        );
        let mut defaults = json!({});
        defaults["endpoint"] = json!(match kind {
            "observable-opentelemetry" => "http://localhost:4318",
            "observable-axiom" => "https://api.axiom.co",
            "observable-zipkin" => "http://localhost:9411/api/v2/spans",
            _ => "",
        });
        if kind == "observable-graylog" {
            defaults["port"] = json!(12201);
            defaults["facility"] = json!("bsb")
        }
        let mut options: Options = serde_json::from_value(crate::config::merge(&defaults, &raw))?;
        ensure!(
            ["production", "production-debug", "development"].contains(&options.mode.as_str()),
            "invalid observable mode"
        );
        if kind == "observable-default" && options.mode != "production" {
            if raw.get("level").is_none() {
                options.level = "debug".into();
            }
            if raw.get("prettyPrint").is_none() {
                options.pretty_print = options.mode == "development";
            }
        }
        ensure!(
            level(&options.level).is_some() && options.base.is_object(),
            "invalid logging level/base"
        );
        ensure!(
            options.max_bytes > 0
                && options.max_bytes <= 1024 * 1024 * 1024
                && (100..=60000).contains(&options.flush_interval_ms)
                && (1..=4096).contains(&options.max_batch_size)
                && options.sampling_rate.is_finite()
                && (0.0..=1.0).contains(&options.sampling_rate),
            "invalid logging/export bounds"
        );
        ensure!(
            ["none", "hourly", "daily"].contains(&options.interval.as_str()),
            "invalid rotation interval"
        );
        let local = matches!(
            kind,
            "observable-default"
                | "observable-logging-file"
                | "observable-pino"
                | "observable-winston"
        );
        let file = if local {
            let path = if kind == "observable-logging-file" {
                &options.path
            } else {
                &options.file_path
            };
            if path.is_empty() {
                None
            } else {
                Some(Mutex::new(crate::telemetry_file::RotatingFile::new(
                    path,
                    options.clone(),
                )?))
            }
        } else {
            None
        };
        let mut network = if kind == "observable-graylog" || kind == "observable-syslog" {
            Some(crate::telemetry_network::Network::new(kind, options.clone()).await?)
        } else {
            None
        };
        if !local && network.is_none() {
            valid_url(
                &options.endpoint,
                kind != "observable-axiom" || options.allow_insecure_http,
            )?;
            if kind == "observable-axiom" {
                ensure!(
                    !options.token.is_empty() && !options.dataset.is_empty(),
                    "Axiom token and dataset required"
                )
            }
        }
        let (sender, mut receiver) = mpsc::channel(4096);
        let result = Arc::new(Self {
            kind: kind.into(),
            options: options.clone(),
            sender,
            stop: CancellationToken::new(),
            closed: AtomicBool::new(false),
            dropped: AtomicU64::new(0),
            worker: AsyncMutex::new(None),
            file,
        });
        if !local {
            let stop = result.stop.clone();
            let kind = kind.to_owned();
            let task = tokio::spawn(async move {
                let mut timer =
                    tokio::time::interval(Duration::from_millis(options.flush_interval_ms));
                timer.tick().await;
                let mut batch = Vec::new();
                loop {
                    tokio::select! {_=stop.cancelled()=>{receiver.close();while let Some(entry)=receiver.recv().await{batch.push(entry);if batch.len()>=options.max_batch_size{flush(&kind,&options,&mut network,std::mem::take(&mut batch)).await;}}flush(&kind,&options,&mut network,batch).await;return},_=timer.tick()=>{},entry=receiver.recv()=>match entry{Some(entry)=>{batch.push(entry);if batch.len()<options.max_batch_size{continue}},None=>return}}
                    flush(&kind, &options, &mut network, std::mem::take(&mut batch)).await;
                }
            });
            *result.worker.lock().await = Some(task)
        }
        Ok(result)
    }
}

pub(super) fn valid_url(raw: &str, allow_http: bool) -> Result<reqwest::Url> {
    let url = reqwest::Url::parse(raw)?;
    ensure!(
        url.has_host()
            && url.username().is_empty()
            && url.password().is_none()
            && url.query().is_none()
            && url.fragment().is_none(),
        "invalid telemetry endpoint"
    );
    ensure!(
        url.scheme() == "https" || (allow_http && url.scheme() == "http"),
        "HTTPS required"
    );
    Ok(url)
}
pub fn redact(mut value: Value, paths: &[String]) -> Value {
    fn visit(value: &mut Value, parts: &[&str]) {
        if parts.is_empty() {
            return;
        }
        match value {
            Value::Object(values) => {
                for (key, child) in values {
                    if parts[0] == "*" || parts[0] == key {
                        if parts.len() == 1 {
                            *child = json!("[REDACTED]")
                        } else {
                            visit(child, &parts[1..])
                        }
                    }
                }
            }
            Value::Array(values) => {
                for (index, child) in values.iter_mut().enumerate() {
                    if parts[0] == "*" || parts[0] == index.to_string() {
                        if parts.len() == 1 {
                            *child = json!("[REDACTED]")
                        } else {
                            visit(child, &parts[1..])
                        }
                    }
                }
            }
            _ => {}
        }
    }
    for path in paths {
        visit(&mut value, &path.split('.').collect::<Vec<_>>())
    }
    value
}
#[async_trait]
impl Observer for Native {
    fn record(&self, value: Value) {
        if self.closed.load(Ordering::SeqCst) {
            return;
        }
        let signal = value["signal"].as_str().unwrap_or_default();
        let local = matches!(
            self.kind.as_str(),
            "observable-default"
                | "observable-logging-file"
                | "observable-pino"
                | "observable-winston"
        );
        if signal == "logs"
            && (!self.options.logs
                || level(value["level"].as_str().unwrap_or("info")).unwrap_or(2)
                    < level(&self.options.level).unwrap()
                || self.kind == "observable-zipkin")
        {
            return;
        }
        if signal == "metrics"
            && (!self.options.metrics
                || local
                || self.kind == "observable-zipkin"
                || self.kind == "observable-graylog"
                || self.kind == "observable-syslog")
        {
            return;
        }
        if signal == "traces" {
            if !self.options.traces
                || local
                || self.kind == "observable-graylog"
                || self.kind == "observable-syslog"
            {
                return;
            }
            let id = value["traceId"].as_str().unwrap_or_default();
            let sample = id.get(..8).and_then(|v| u32::from_str_radix(v, 16).ok());
            if sample
                .is_none_or(|sample| sample as f64 / 4294967296.0 >= self.options.sampling_rate)
            {
                return;
            }
        }
        let mut value = redact(
            crate::config::merge(&self.options.base, &value),
            &self.options.redact,
        );
        if let (Some(message), Some(meta)) = (value["message"].as_str(), value["meta"].as_object())
        {
            let mut message = message.to_owned();
            for (key, item) in meta {
                message = message.replace(
                    &format!("{{{key}}}"),
                    item.as_str()
                        .map(str::to_owned)
                        .unwrap_or_else(|| item.to_string())
                        .as_str(),
                )
            }
            value["message"] = json!(message)
        }
        if self.kind == "observable-pino" {
            value["level"] =
                json!((level(value["level"].as_str().unwrap_or("info")).unwrap_or(2) + 1) * 10);
            value["msg"] = value["message"].take();
            value.as_object_mut().unwrap().remove("message");
        }
        let bytes = match serde_json::to_vec(&value) {
            Ok(bytes) if bytes.len() <= 64 * 1024 => bytes,
            _ => {
                eprintln!("BSB telemetry entry exceeds 64 KiB");
                return;
            }
        };
        if local {
            if self.kind != "observable-logging-file" {
                if self.options.pretty_print {
                    println!(
                        "{} [{}] {}: {}",
                        value["timestamp"].as_str().unwrap_or_default(),
                        value["level"],
                        value["plugin"].as_str().unwrap_or_default(),
                        value
                            .get("message")
                            .or_else(|| value.get("msg"))
                            .and_then(Value::as_str)
                            .unwrap_or_default()
                    )
                } else {
                    println!("{value}")
                }
            }
            if let Some(file) = &self.file {
                if file.lock().unwrap().write(&bytes).is_err() {
                    eprintln!("BSB file logging failed")
                }
            }
        } else if self.sender.try_send(value).is_err() {
            let dropped = self.dropped.fetch_add(1, Ordering::Relaxed) + 1;
            if dropped % 1000 == 1 {
                eprintln!("BSB telemetry queue full; dropped {dropped} entries")
            }
        }
    }
    async fn shutdown(&self) -> Result<()> {
        if self.closed.swap(true, Ordering::SeqCst) {
            return Ok(());
        }
        self.stop.cancel();
        if let Some(mut worker) = self.worker.lock().await.take() {
            if tokio::time::timeout(Duration::from_secs(10), &mut worker)
                .await
                .is_err()
            {
                worker.abort();
                let _ = worker.await;
            }
        }
        if let Some(file) = &self.file {
            file.lock().unwrap().close()?;
        }
        Ok(())
    }
}
async fn flush(
    kind: &str,
    options: &Options,
    network: &mut Option<crate::telemetry_network::Network>,
    batch: Vec<Value>,
) {
    if batch.is_empty() {
        return;
    }
    let result = if let Some(network) = network {
        network.export(batch).await
    } else {
        export(kind, options, batch).await
    };
    if result.is_err() {
        eprintln!("BSB {kind}: telemetry export failed")
    }
}
pub(super) async fn post(
    endpoint: &str,
    body: Value,
    headers: &BTreeMap<String, String>,
) -> Result<()> {
    let mut map = HeaderMap::new();
    for (key, value) in headers {
        map.insert(
            HeaderName::from_bytes(key.as_bytes())?,
            HeaderValue::from_str(value)?,
        );
    }
    let url = valid_url(endpoint, true)?;
    let client = http::client(Duration::from_secs(5))?;
    for attempt in 0..3 {
        match http::request(
            &client,
            reqwest::Method::POST,
            url.clone(),
            map.clone(),
            Some(&body),
            64 * 1024 * 1024,
        )
        .await
        {
            Ok(value) => {
                if let Some(partial) = value["partialSuccess"].as_object() {
                    for (key, value) in partial {
                        ensure!(
                            if key == "errorMessage" {
                                value == ""
                            } else {
                                value == 0 || value == "0"
                            },
                            "collector partially rejected telemetry"
                        )
                    }
                }
                ensure!(
                    value.get("failed").is_none_or(|v| v.as_u64() == Some(0)),
                    "collector rejected telemetry"
                );
                return Ok(());
            }
            Err(error) => {
                if attempt == 2 || !http::retryable(&error) {
                    return Err(error);
                }
            }
        }
        tokio::time::sleep(Duration::from_millis((attempt + 1) * 250)).await;
    }
    unreachable!()
}
fn any_value(value: &Value) -> Value {
    match value {
        Value::Bool(v) => json!({"boolValue":v}),
        Value::Number(v) => {
            if v.is_i64() || v.is_u64() {
                json!({"intValue":v.to_string()})
            } else {
                json!({"doubleValue":v})
            }
        }
        Value::Array(v) => {
            json!({"arrayValue":{"values":v.iter().map(any_value).collect::<Vec<_>>()}})
        }
        Value::Object(_) => json!({"kvlistValue":{"values":attributes(value)}}),
        _ => {
            json!({"stringValue":value.as_str().map(str::to_owned).unwrap_or_else(||value.to_string())})
        }
    }
}
fn attributes(value: &Value) -> Vec<Value> {
    value
        .as_object()
        .map(|v| {
            v.iter()
                .map(|(key, value)| json!({"key":key,"value":any_value(value)}))
                .collect()
        })
        .unwrap_or_default()
}
fn otlp(signal: &str, options: &Options, entries: Vec<Value>) -> Value {
    let mut resource = json!(options.resource_attributes);
    resource["service.name"] = json!(options.service_name);
    if !options.service_version.is_empty() {
        resource["service.version"] = json!(options.service_version)
    }
    let mut grouped: BTreeMap<String, Vec<Value>> = BTreeMap::new();
    let entries = if signal == "metrics" {
        let mut latest = BTreeMap::new();
        for entry in entries {
            latest.insert(
                format!("{}\0{}\0{}", entry["plugin"], entry["name"], entry["kind"]),
                entry,
            );
        }
        latest.into_values().collect()
    } else {
        entries
    };
    for entry in entries {
        let item = match signal {
            "logs" => {
                let timestamp = entry["timestamp"]
                    .as_str()
                    .and_then(|s| chrono::DateTime::parse_from_rfc3339(s).ok())
                    .and_then(|v| v.timestamp_nanos_opt())
                    .unwrap_or_default();
                let severity = entry["level"].as_str().unwrap_or("info");
                json!({"timeUnixNano":timestamp.to_string(),"severityNumber":1+level(severity).unwrap_or(2)*4,"severityText":severity.to_uppercase(),"body":any_value(&entry["message"]),"attributes":attributes(&entry["meta"]),"traceId":entry["traceId"],"spanId":entry["spanId"]})
            }
            "traces" => {
                json!({"traceId":entry["traceId"],"spanId":entry["spanId"],"parentSpanId":entry["parentSpanId"],"name":entry["name"],"kind":1,"startTimeUnixNano":entry["startedNs"],"endTimeUnixNano":entry["endedNs"],"attributes":attributes(&entry["attributes"]),"status":if entry["error"].as_str().is_some_and(|v|!v.is_empty()){json!({"code":2,"message":entry["error"]})}else{json!({"code":0})}})
            }
            _ => {
                let mut point = json!({"timeUnixNano":entry["timestampNs"],"startTimeUnixNano":entry["startedNs"]});
                let mut metric = json!({"name":entry["name"],"description":entry["description"],"unit":entry["unit"]});
                match entry["kind"].as_str() {
                    Some("counter") => {
                        point["asInt"] = json!(entry["value"].to_string());
                        metric["sum"] = json!({"aggregationTemporality":2,"isMonotonic":true,"dataPoints":[point]})
                    }
                    Some("histogram") => {
                        point["count"] = json!(entry["count"].to_string());
                        point["sum"] = entry["sum"].clone();
                        point["bucketCounts"] = json!([entry["count"].to_string()]);
                        point["explicitBounds"] = json!([]);
                        metric["histogram"] =
                            json!({"aggregationTemporality":2,"dataPoints":[point]})
                    }
                    _ => {
                        point["asDouble"] = entry["value"].clone();
                        metric["gauge"] = json!({"dataPoints":[point]})
                    }
                }
                metric
            }
        };
        grouped
            .entry(entry["plugin"].as_str().unwrap_or("bsb").into())
            .or_default()
            .push(item);
    }
    let (suffix, field) = match signal {
        "logs" => ("Logs", "logRecords"),
        "traces" => ("Spans", "spans"),
        _ => ("Metrics", "metrics"),
    };
    let scopes: Vec<_> = grouped
        .into_iter()
        .map(|(name, items)| json!({"scope":{"name":name},field:items}))
        .collect();
    json!({format!("resource{suffix}"):[{"resource":{"attributes":attributes(&resource)},format!("scope{suffix}"):scopes}]})
}
async fn export(kind: &str, options: &Options, batch: Vec<Value>) -> Result<()> {
    let endpoint = options.endpoint.trim_end_matches('/');
    let mut headers = options.headers.clone();
    if kind == "observable-zipkin" {
        let spans:Vec<_>=batch.into_iter().filter(|v|v["signal"]=="traces").map(|entry|{
            let start=entry["startedNs"].as_str().and_then(|v|v.parse::<i64>().ok()).unwrap_or_default();
            let end=entry["endedNs"].as_str().and_then(|v|v.parse::<i64>().ok()).unwrap_or(start);
            let mut tags:BTreeMap<String,String>=entry["attributes"].as_object().into_iter().flatten().map(|(key,value)|(key.clone(),value.as_str().map(str::to_owned).unwrap_or_else(||value.to_string()))).collect();
            if let Some(error)=entry["error"].as_str(){tags.insert("error".into(),error.into());}
            json!({"traceId":entry["traceId"],"id":entry["spanId"],"parentId":entry["parentSpanId"],"name":entry["name"],"timestamp":start/1000,"duration":((end-start)/1000).max(1),"tags":tags,"localEndpoint":{"serviceName":options.service_name}})
        }).collect();
        return post(endpoint, json!(spans), &headers).await;
    }
    if kind == "observable-axiom" {
        headers.insert("Authorization".into(), format!("Bearer {}", options.token));
        headers.insert("X-Axiom-Dataset".into(), options.dataset.clone());
        if !options.org_id.is_empty() {
            headers.insert("X-Axiom-Org-Id".into(), options.org_id.clone());
        }
        let events: Vec<_> = batch
            .iter()
            .filter(|v| v["signal"] != "traces")
            .map(|v| {
                let mut entry = v.clone();
                entry["_time"] = entry
                    .get("timestamp")
                    .cloned()
                    .unwrap_or_else(|| json!(chrono::Utc::now().to_rfc3339()));
                entry["service"] = json!(options.service_name);
                entry
            })
            .collect();
        if !events.is_empty() {
            let mut url = valid_url(endpoint, options.allow_insecure_http)?;
            url.path_segments_mut()
                .unwrap()
                .extend(["v1", "datasets", &options.dataset, "ingest"]);
            post(url.as_str(), json!(events), &headers).await?
        }
    }
    for signal in ["logs", "metrics", "traces"] {
        if kind == "observable-axiom" && signal != "traces" {
            continue;
        }
        let entries: Vec<_> = batch
            .iter()
            .filter(|v| v["signal"] == signal)
            .cloned()
            .collect();
        if !entries.is_empty() {
            post(
                &format!("{endpoint}/v1/{signal}"),
                otlp(signal, options, entries),
                &headers,
            )
            .await?
        }
    }
    Ok(())
}
